numba/__init__.py

```diff
@@ -58,6 +59,90 @@
+    'numba.compiler': 'numba.core.compiler',
+    'numba.compiler_lock': 'numba.core.compiler_lock',
+    'numba.compiler_machinery': 'numba.core.compiler_machinery',
+    'numba.consts': 'numba.core.consts',
+    'numba.controlflow': 'numba.core.controlflow',
+    'numba.dataflow': 'numba.core.dataflow',
+    'numba.debuginfo': 'numba.core.debuginfo',
+    'numba.decorators': 'numba.core.decorators',
+    'numba.dictobject': 'numba.typed.dictobject',
+    'numba.dispatcher': 'numba.core.dispatcher',
+    'numba.entrypoints': 'numba.core.entrypoints',
+    'numba.funcdesc': 'numba.core.funcdesc',
+    'numba.generators': 'numba.core.generators',
+    'numba.inline_closurecall': 'numba.core.inline_closurecall',
+    'numba.interpreter': 'numba.core.interpreter',
+    'numba.ir': 'numba.core.ir',
+    'numba.ir_utils': 'numba.core.ir_utils',
+    'numba.itanium_mangler': 'numba.core.itanium_mangler',
+    'numba.listobject': 'numba.typed.listobject',
+    'numba.lowering': 'numba.core.lowering',
+    'numba.npdatetime': 'numba.np.npdatetime',
+    'numba.object_mode_passes': 'numba.core.object_mode_passes',
+    'numba.parfor': 'numba.parfors.parfor',
+    'numba.postproc': 'numba.core.postproc',
+    'numba.pylowering': 'numba.core.pylowering',
+    'numba.pythonapi': 'numba.core.pythonapi',
+    'numba.rewrites': 'numba.core.rewrites',
+    'numba.runtime': 'numba.core.runtime',
+    'numba.serialize': 'numba.core.serialize',
+    'numba.sigutils': 'numba.core.sigutils',
+    'numba.stencilparfor': 'numba.stencils.stencilparfor',
+    'numba.tracing': 'numba.core.tracing',
+    'numba.transforms': 'numba.core.transforms',
+    'numba.typeconv': 'numba.core.typeconv',
+    'numba.withcontexts': 'numba.core.withcontexts',
+    'numba.typed_passes': 'numba.core.typed_passes',
+    'numba.untyped_passes': 'numba.core.untyped_passes',
+    'numba.utils': 'numba.core.utils',
+    'numba.unicode_support': 'numba.cpython.unicode_support',
+    'numba.unicode': 'numba.cpython.unicode',
+    'numba.typing': 'numba.core.typing',
+    'numba.typeinfer': 'numba.core.typeinfer',
+    'numba.typedobjectutils': 'numba.typed.typedobjectutils',
+}
+
+if sys.version_info < (3, 7):
+    with warnings.catch_warnings():
+        # Ignore warnings from making top-level alias.
+        warnings.simplefilter(
+            "ignore", category=errors.NumbaDeprecationWarning,
+        )
+        for _old_mod, _new_mod in _auto_import_submodules.items():
+            importlib.import_module(_old_mod)
+else:
+    def __getattr__(attr):
+        submodule_name = 'numba.{}'.format(attr)
+        try:
+            new_name = _auto_import_submodules[submodule_name]
+        except KeyError:
+            raise AttributeError(
+                f"module {__name__!r} has no attribute {attr!r}",
+            )
+        else:
+            errors.deprecate_moved_module(submodule_name, new_name)
+            return importlib.import_module(new_name)
```

Comment:
Another potentially scary option would be:
```python
class _NumbaModule(types.ModuleType):
    def __getattr__(self, attr):
        ...  # as here

...

# as the very last step
_new_module = _NumbaModule(__name__)
_new_module.__dict__.update(globals()
sys.modules[__name__] = _new_module
```

This would make the warning appear on python 3.6 too

Reply:
Which would remove the need for the test changes here, I think

Reply:
It does seem very scary to patch the toplevel module